terrustaria/src/map.rs

```diff
@@ -0,0 +1,182 @@
+    texture_index: TileTextureIndex,
+    valid: Box<dyn Fn(u32, u32) -> bool>,
+}
+
+impl TileType {
+    fn new(
+        name: String,
+        occurrence_prob: f32,
+        texture_index: TileTextureIndex,
+        valid: impl Fn(u32, u32) -> bool + 'static,
+    ) -> Self {
+        Self {
+            name,
+            occurrence_prob,
+            texture_index,
+            valid: Box::new(valid),
+        }
+    }
+}
+
+fn init_tile_types() -> Vec<TileType> {
+    vec![
+        TileType::new(String::from("Grass"), 0.6, TileTextureIndex(0), |_, _| true),
+        TileType::new(String::from("Stone"), 0.3, TileTextureIndex(3), |_, y| {
+            y < 12
+        }),
+        TileType::new(String::from("Water"), 0.1, TileTextureIndex(1), |x, y| {
+            y < 20 && x % 2 == 0
+        }),
+    ]
+}
+
+fn get_random_tile_type(tile_types: &[TileType]) -> usize {
+    let mut val = thread_rng().gen();
+    for (i, tile_type) in tile_types.iter().enumerate() {
+        let prob = tile_type.occurrence_prob;
+        if prob >= val {
+            return i;
+        } else {
+            val -= prob;
+        }
+    }
+    0
+}
+
+// Fills the tilemap with set texture_id, does not fill building area.
+fn fill_tilemap_with_set_structure_id(
+    texture_index: TileTextureIndex,
+    tilemap_id: TilemapId,
+    commands: &mut Commands,
+    tile_storage: &mut TileStorage,
+    map_name: &str,
+) {
+    for x in 0..MAP_SIZE.x {
+        for y in 0..MAP_SIZE.y - BUILDING_HEIGHT {
+            let tile_pos = TilePos { x, y };
+            let tile_entity = commands
+                .spawn(TileBundle {
+                    position: tile_pos,
+                    tilemap_id,
+                    texture_index,
+                    ..Default::default()
+                })
+                .insert(Name::new(format!("{map_name}Tile({x},{y})")))
+                .id();
+            tile_storage.set(&tile_pos, tile_entity);
+        }
+    }
+}
+
+// Fills randomly tilemap with colliders and textures, does not fill building area.
+fn fill_tilemap_randomly_with_colliders(
+    tilemap_id: TilemapId,
+    commands: &mut Commands,
+    tile_storage: &mut TileStorage,
+    map_name: &str,
+) {
+    let tile_types = init_tile_types();
+
+    for x in 0..MAP_SIZE.x {
+        for y in 0..MAP_SIZE.y - BUILDING_HEIGHT {
+            let tile_pos = TilePos { x, y };
+            let mut idx: usize;
+            loop {
+                idx = get_random_tile_type(&tile_types);
+                if (tile_types[idx].valid)(x, y) {
+                    break;
+                }
+            }
+
+            let tile_entity = commands
+                .spawn(TileBundle {
+                    position: tile_pos,
+                    tilemap_id,
+                    texture_index: tile_types[idx].texture_index,
+                    ..Default::default()
+                })
+                .insert(Name::new(format!("{map_name}Tile({x},{y})")))
```

Comment:
a jeśli chcecie użyć nazwy mapy i jednocześnie uzywać enuma zamiast stringa, to wystarczy zrobić #[derive(Debug)] na enumie i tutaj wyprintować nazwę przy użyciu formattera debugowego... albo samemu zaimplementować Display dla enuma i zostawić jak jest